### ISSUE TITLE
Template transformation can cause default parameters to be interpreted as numbers

### ISSUE THREAD

**Issue by RossCasey** [aws/aws-sam-cli#1094]
### Description

When `sam build` transforms the template file, default parameters with quotes around them appear in the output file without quotes. This results in strings which contain numbers being interpreted as numbers. 

### Steps to reproduce

Execute `sam build` on the following template:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'

Parameters:
  MyParam:
    Type: String
    Default: "012345678912"

Resources:
  SSM:
    Type: AWS::SSM::Parameter
    Properties:
      Name: MyParam
      Value: !Ref MyParam
      Type: String
```

### Observed result

This results in the following in `.aws-sam/build/template.yaml`:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Parameters:
  MyParam:
    Default: 012345678912
    Type: String
Resources:
  SSM:
    Properties:
      Name: MyParam
      Type: String
      Value:
        Ref: MyParam
    Type: AWS::SSM::Parameter
Transform: AWS::Serverless-2016-10-31
```

This can result in errors such as the following within Cloudformation:
```
1 validation error detected: Value 'arn:aws:sns:us-east-1:1.234567890E10:my-topic' at 'sourceArn' failed to satisfy constraint: Member must satisfy regular expression pattern: arn:(aws[a-zA-Z0-9-]*):([a-zA-Z0-9\\-])+:([a-z]{2}((-gov)|(-iso(b?)))?-[a-z]+-\\d{1})?:(\\d{12})?:(.*)
```

Where the value has been interpreted as a number.

### Expected result

The outputted template should also contain the quotes:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Parameters:
  MyParam:
    Default: "012345678912"
    Type: String
Resources:
  SSM:
    Properties:
      Name: MyParam
      Type: String
      Value:
        Ref: MyParam
    Type: AWS::SSM::Parameter
Transform: AWS::Serverless-2016-10-31
```

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: macOS Mojave
2. `sam --version`: SAM CLI, version 0.14.2

Debug output for running `sam build`:

Debug output running above command:

```
2019-04-01 16:00:37 Changing event name from creating-client-class.iot-data to creating-client-class.iot-data-plane
2019-04-01 16:00:37 Changing event name from before-call.apigateway to before-call.api-gateway
2019-04-01 16:00:37 Changing event name from request-created.machinelearning.Predict to request-created.machine-learning.Predict
2019-04-01 16:00:37 Changing event name from before-parameter-build.autoscaling.CreateLaunchConfiguration to before-parameter-build.auto-scaling.CreateLaunchConfiguration
2019-04-01 16:00:37 Changing event name from before-parameter-build.route53 to before-parameter-build.route-53
2019-04-01 16:00:37 Changing event name from request-created.cloudsearchdomain.Search to request-created.cloudsearch-domain.Search
2019-04-01 16:00:37 Changing event name from docs.*.autoscaling.CreateLaunchConfiguration.complete-section to docs.*.auto-scaling.CreateLaunchConfiguration.complete-section
2019-04-01 16:00:37 Changing event name from before-parameter-build.logs.CreateExportTask to before-parameter-build.cloudwatch-logs.CreateExportTask
2019-04-01 16:00:37 Changing event name from docs.*.logs.CreateExportTask.complete-section to docs.*.cloudwatch-logs.CreateExportTask.complete-section
2019-04-01 16:00:37 Changing event name from before-parameter-build.cloudsearchdomain.Search to before-parameter-build.cloudsearch-domain.Search
2019-04-01 16:00:37 Changing event name from docs.*.cloudsearchdomain.Search.complete-section to docs.*.cloudsearch-domain.Search.complete-section
2019-04-01 16:00:37 Setting config variable for profile to 'profile-name'
2019-04-01 16:00:37 Changing event name from creating-client-class.iot-data to creating-client-class.iot-data-plane
2019-04-01 16:00:37 Changing event name from before-call.apigateway to before-call.api-gateway
2019-04-01 16:00:37 Changing event name from request-created.machinelearning.Predict to request-created.machine-learning.Predict
2019-04-01 16:00:37 Changing event name from before-parameter-build.autoscaling.CreateLaunchConfiguration to before-parameter-build.auto-scaling.CreateLaunchConfiguration
2019-04-01 16:00:37 Changing event name from before-parameter-build.route53 to before-parameter-build.route-53
2019-04-01 16:00:37 Changing event name from request-created.cloudsearchdomain.Search to request-created.cloudsearch-domain.Search
2019-04-01 16:00:37 Changing event name from docs.*.autoscaling.CreateLaunchConfiguration.complete-section to docs.*.auto-scaling.CreateLaunchConfiguration.complete-section
2019-04-01 16:00:38 Changing event name from before-parameter-build.logs.CreateExportTask to before-parameter-build.cloudwatch-logs.CreateExportTask
2019-04-01 16:00:38 Changing event name from docs.*.logs.CreateExportTask.complete-section to docs.*.cloudwatch-logs.CreateExportTask.complete-section
2019-04-01 16:00:38 Changing event name from before-parameter-build.cloudsearchdomain.Search to before-parameter-build.cloudsearch-domain.Search
2019-04-01 16:00:38 Changing event name from docs.*.cloudsearchdomain.Search.complete-section to docs.*.cloudsearch-domain.Search.complete-section
2019-04-01 16:00:38 Setting config variable for profile to 'profile-name'
2019-04-01 16:00:38 Setting config variable for region to 'eu-west-1'
2019-04-01 16:00:38 Using SAM Template at path
2019-04-01 16:00:38 'build' command is called
2019-04-01 16:00:38 Collected default values for parameters: {'MyParam': '012345678912'}
2019-04-01 16:00:38 1 resources found in the template
```

**Comment by meh485** [comment]
I'm having the same problem running
SAM CLI, version 0.17.0

The weird part is, samcli puts some account numbers in single quotes but not others. See example
Yaml template before using sam build:
`
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'

Mappings:
  ELBAccountPrincipals:
    us-east-1:
      "AccountId": "127311923021"
    us-east-2:
      "AccountId": "033677994240"
    us-west-1:
      "AccountId": "027434742980"      
    us-west-2:
      "AccountId": "797873946194"
    ca-central-1:
      "AccountId": "985666609251"
    eu-central-1:
      "AccountId": "054676820928"
    eu-west-1:
      "AccountId": "156460612806"
    eu-west-2:
      "AccountId": "652711504416"      
    eu-west-3:
      "AccountId": "009996457667"
    eu-north-1:
      "AccountId": "897822967062"      
    ap-northeast-1:
      "AccountId": "582318560864"
    ap-northeast-2:
      "AccountId": "600734575887"
    ap-northeast-3:
      "AccountId": "383597477331"
    ap-southeast-1:
      "AccountId": "114774131450"
    ap-southeast-2:
      "AccountId": "783225319266"    
    ap-south-1:
      "AccountId": "718504428378"     
    sa-east-1:
      "AccountId": "507241528517"  
    us-gov-west-1:
      "AccountId": "048591011584"
    us-gov-east-1:
      "AccountId": "190560391635"              
    cn-northwest-1:
      "AccountId": "037604701340"
    cn-north-1:
      "AccountId": "638102146993"
`

Output from sam build
`
AWSTemplateFormatVersion: '2010-09-09'
Transform:
- AWS::Serverless-2016-10-31
Mappings:
  ELBAccountPrincipals:
    ap-northeast-1:
      AccountId: '582318560864'
    ap-northeast-2:
      AccountId: '600734575887'
    ap-northeast-3:
      AccountId: '383597477331'
    ap-south-1:
      AccountId: '718504428378'
    ap-southeast-1:
      AccountId: '114774131450'
    ap-southeast-2:
      AccountId: '783225319266'
    ca-central-1:
      AccountId: '985666609251'
    cn-north-1:
      AccountId: '638102146993'
    cn-northwest-1:
      AccountId: '037604701340'
    eu-central-1:
      AccountId: 054676820928
    eu-north-1:
      AccountId: '897822967062'
    eu-west-1:
      AccountId: '156460612806'
    eu-west-2:
      AccountId: '652711504416'
    eu-west-3:
      AccountId: 009996457667
    sa-east-1:
      AccountId: '507241528517'
    us-east-1:
      AccountId: '127311923021'
    us-east-2:
      AccountId: 033677994240
    us-gov-east-1:
      AccountId: '190560391635'
    us-gov-west-1:
      AccountId: 048591011584
    us-west-1:
      AccountId: 027434742980
    us-west-2:
      AccountId: '797873946194'
`

You can see sam puts quotes around 
    cn-northwest-1:
      AccountId: '037604701340'
but not any others that start with 0
    us-west-1:
      AccountId: 027434742980

**Comment by nowak-ninja** [comment]
Same for me, CloudFormation will fail when trying to use account IDs with leading zeroes.

**Comment by moelasmar** [comment]
I can confirm that this issue is fixed in SAM CLI version 1.35.0
